utilities/checkpoint/checkpoint_impl.cc

```diff
@@ -588,6 +645,37 @@ Status CheckpointImpl::ExportFilesInMetaData(
 
   return s;
 }
+
+Status CheckpointImpl::CopyOptionsFile(const std::string& src_file,
+                                       const std::string& target_file,
+                                       const std::string& db_log_dir,
+                                       const std::string& wal_dir) {
+  Status s;
+  DBOptions src_db_options;
+  std::vector<ColumnFamilyDescriptor> src_cf_descs;
+  s = LoadOptionsFromFile(ConfigOptions(), src_file, &src_db_options,
+                          &src_cf_descs);
+  if (!s.ok()) {
+    return s;
+  }
+
+  // Override these 2 options
+  src_db_options.db_log_dir = db_log_dir;
+  src_db_options.wal_dir = wal_dir;
+
+  std::vector<std::string> src_cf_names;
+  std::vector<ColumnFamilyOptions> src_cf_opts;
+  src_cf_names.reserve(src_cf_descs.size());
+  src_cf_opts.reserve(src_cf_descs.size());
+  for (ColumnFamilyDescriptor desc : src_cf_descs) {
+    src_cf_names.push_back(desc.name);
+    src_cf_opts.push_back(desc.options);
+  }
+
+  return PersistRocksDBOptions(src_db_options, src_cf_names, src_cf_opts,
+                               target_file, db_->GetFileSystem());
```

Comment:
@ajkr I am not sure I understand your concern regarding custom objects and how it relates to this PR.  The problem of re-creating custom objects already exists and is being addressed by the plug-in/customizable PRs.  As more objects are registered with the ObjectRegistry/Plugin infrastructure, it should be possible to re-create almost all of the settings from the file.  I am not sure I understand how this PR makes anything worse regarding custom objects.  

There are options I can see that probably should not be serialized to the OPTIONS file (like "create_if_missing") and options that need to be altered on a checkpoint (not only wal_dir and db_log_dir, but also db_path and cf_path).  In a checkpoint, those options need to be altered.  This PR is a first-pass at attempting to update the wal_dir and db_log_dir in a checkpoint. Outside of adding the new parameters, I can see value in updating the OPTIONS.

Reply:
> The problem of re-creating custom objects already exists

It doesn't exist for a running DB that calls PersistRocksDBOptions(), but it does exist for this new client of PersistRocksDBOptions()

Reply:
> > The problem of re-creating custom objects already exists
> 
> It doesn't exist for a running DB that calls PersistRocksDBOptions(), but it does exist for this new client of PersistRocksDBOptions()
> 
> edit: internal task link is T96547033

Prior to this change, an OPTIONS file was already written/copied for a checkpoint:
. /ldb --version
ldb from RocksDB 6.22.3
./ldb --db=../ldb-db --try_load_options checkpoint --checkpoint_dir=../ldb-cp --ignore_unknown_options
s -l ../ldb-cp
total 20
-rw-r--r-- 2 rocksdb rocksdb  972 Jul 27 17:21 000009.sst
-rw-r--r-- 1 rocksdb rocksdb   16 Jul 27 17:32 CURRENT
-rw-r--r-- 1 rocksdb rocksdb  108 Jul 27 17:32 MANIFEST-000020
-rw-r--r-- 1 rocksdb rocksdb 6222 Jul 27 17:32 OPTIONS-000023

The difference is that this change attempts to write an OPTIONS file with the correct values for wal_dir.

Reply:
Anyways we need to either fix T96547033 or rollback soon since it is broken on master. I suggest rollback as the bug seems difficult to fix.

Reply:
It seems like it would be useful to have another shallower abstraction of deserialized options: able to mutate in a "structured" way but no resolved bindings. Or perhaps just some way to specify some overrides and read+write the file applying those overrides (without attempting to resolve any bindings).